File: tools/maze.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# ============================================================================ #
# author: Ryotaro Onuki (kerikun11+github@gmail.com)
# description: a maze module includes Maze class
# usage: $ python maze.py mazefile.maze
# python version >= 3.8
# ============================================================================ #
import sys
import numpy as np


class Maze:
    """
    a maze class includes wall data, start indexes, goal indexes
    """
    # constants
    East, North, West, South = range(4)

    def __init__(self, size=32, start=[], goals=[]):
        """
        construct a maze object with a maze size

        Returns
        -------
        None
        """
        self.size = size  # number of cells on one side of the maze
        # the number of cells in the maze: x * y
        self.cell_index_size = size * size
        # the number of walls to save: x * y * z
        self.wall_index_size = size * size * 2
        # wall data; wall states and known flags
        self.walls = np.zeros(self.wall_index_size, dtype=bool)
        self.knowns = np.zeros(self.wall_index_size, dtype=bool)
        # start and goal cells
        self.start = start
        self.goals = goals

    @classmethod
    def uniquify(cls, x, y, d):
        """
        returns a unique coordinates of a wall without redundancy on both sides of the wall

        Returns
        -------
        int x, int y, int z, int d
        """
        if d == cls.East:
            x, y, z, d = x, y, 0, cls.East
        elif d == cls.North:
            x, y, z, d = x, y, 1, cls.North
        elif d == cls.West:
            x, y, z, d = x-1, y, 0, cls.East
        elif d == cls.South:
            x, y, z, d = x, y-1, 1, cls.North
        return x, y, z, d

    def get_wall_index(self, x, y, z):
        """
        get a unique and sequential index of a wall inside of the maze

        Returns
        -------
        int index
        """
        if not self.is_inside_of_field(x, y, z):
            raise ValueError("out of field!")
        return x + y * self.size + z * self.size * self.size

    def get_cell_index(self, x, y):
        """
        get a unique and sequential index of a cell inside of the maze

        Returns
        -------
        int index
        """
        if not self.is_inside_of_field(x, y):
            raise ValueError("out of field!")
        return y * self.size + x

    def is_inside_of_field(self, x, y, z=None):
        """
        determine if the wall or cell is inside of the field

        Returns
        -------
        bool result
        """
        s = self.size  # maze size
        if z == None:
            return x >= 0 and y >= 0 and x < s and y < s  # cell
        else:
            return x >= 0 and y >= 0 and x < s+z-1 and y < s-z  # wall

    def wall(self, x, y, d, new_state=None, new_known=None):
        """
        get or update a wall flag, and optionally update a known flag

        Returns
        -------
        bool flag
        """
        x, y, z, d = self.uniquify(x, y, d)
        # If it is out of the field, the wall is assumed to exist.
        if not self.is_inside_of_field(x, y, z):
            return True
        i = self.get_wall_index(x, y, z)
        if new_state != None:
            self.walls[i] = new_state
        if new_known != None:
            self.knowns[i] = new_known
        return self.walls[i]

    def known(self, x, y, d, new_known=None):
        """
        get or update a known flag of a wall

        Returns
        -------
        bool flag
        """
        x, y, z, d = self.uniquify(x, y, d)
        # If it is out of the field, the wall is assumed to be known.
        if not self.is_inside_of_field(x, y, z):
            return True
        i = self.get_wall_index(x, y, z)
        if new_known != None:
            self.knowns[i] = new_known
        return self.knowns[i]

    def __str__(self):
        """
        show information of the maze

        Returns
        -------
        string
        """
        return f'size: {self.size}x{self.size}' + '\n' + \
            'start: ' + ', '.join([f'({x}, {y})' for x, y in self.start]) + '\n' + \
            'goals: ' + ', '.join([f'({x}, {y})' for x, y in self.goals])

    @staticmethod
    def parse_maze_string(file):
        """
        parse a maze string from file and construct a maze object

        Returns
        -------
        Maze object
        """
        lines = file.readlines()
        maze_size = max(len(lines)//2, len(lines[0])//4)
        maze = Maze(maze_size)  # construct a maze object
        for i, line in enumerate(reversed(lines)):
            line = line.rstrip()  # remove \n
            # +---+---+---+---+
            if i % 2 == 0:
                for j, c in enumerate(line[2:: 4]):
                    if c == '-':
                        maze.wall(j, i//2, Maze.South, True, new_known=True)
                    elif c == ' ':
                        maze.wall(j, i//2, Maze.South, False, new_known=True)
            # |   |   | G |   |
            else:
                for j, c in enumerate(line[0:: 4]):
                    if c == '|':
                        maze.wall(j, i//2, Maze.West, True, new_known=True)
                    elif c == ' ':
                        maze.wall(j, i//2, Maze.West, False, new_known=True)
                for j, c in enumerate(line[2:: 4]):
                    if c == 'S':
                        maze.start.append([j, i//2])
                    if c == 'G':
                        maze.goals.append([j, i//2])
        return maze

    def get_maze_string(self):
        """
        generate a maze string to be saved in text format

        Returns
        -------
        string
        """
        res = ''  # result string
        for y in reversed(range(-1, self.size)):
            # +---+---+---+---+
            res += '+'  # first pillar
            for x in range(self.size):
                # horizontal wall
                if not self.known(x, y, Maze.North):
                    res += ' . '
                elif self.wall(x, y, Maze.North):
                    res += '---'
                else:
                    res += '   '
                res += '+'  # pillar
            res += '\n'
            # |   |   | G |   |
            if y == -1:
                break
            res += '|'  # first wall
            for x in range(self.size):
                # cell space
                if [x, y] in self.start:
                    res += ' S '
                elif [x, y] in self.goals:
                    res += ' G '
                else:
                    res += '   '
                # vertical wall
                if not self.known(x, y, Maze.East):
                    res += '.'
                elif self.wall(x, y, Maze.East):
                    res += '|'
                else:
                    res += ' '
            res += '\n'
        return res


# ============================================================================ #
# example
if __name__ == "__main__":
    # check arguments
    if len(sys.argv) < 2:
        print('please specify a maze file.')
        sys.exit(1)

    # set filepath
    filepath = sys.argv[1]

    # read maze file
    with open(filepath, 'r') as file:
        maze = Maze.parse_maze_string(file)

    # show info
    print(maze)
    print(maze.get_maze_string())

File: docs/maze.js
/**
 * @description maze class
 * @author Ryotaro Onuki <kerikun11+github@gmail.com>
 * @date 2021.01.17
 */
"use strict";

export default class Maze {
    static East = 0;
    static North = 1;
    static West = 2;
    static South = 3;
    constructor(size = 32, start = [0, 0], goals = []) {
        this.size = size;
        this.cell_index_size = size * size;
        this.wall_index_size = size * size * 2;
        this.walls = Array(this.wall_index_size).fill(false);
        this.knowns = Array(this.wall_index_size).fill(false);
        this.start = start;
        this.goals = goals;
    }
    static uniquify(x, y, d) {
        switch (d) {
            case Maze.East: return [x, y, 0, d];
            case Maze.North: return [x, y, 1, d];
            case Maze.West: return [x - 1, y, 0, Maze.East];
            case Maze.South: return [x, y - 1, 1, Maze.North];
        }
    }
    get_wall_index(x, y, z) {
        if (!this.is_inside_of_field(x, y, z))
            throw "out of field!";
        return x + y * this.size + z * this.size * this.size;
    }
    get_cell_index(x, y) {
        if (!this.is_inside_of_field(x, y))
            throw "out of field!";
        return x + y * this.size;
    }
    is_inside_of_field(x, y, z = void 0) {
        let s = this.size;
        if (z === void 0)
            return x >= 0 && y >= 0 && x < s && y < s;
        else
            return x >= 0 && y >= 0 && x < s + z - 1 && y < s - z;
    }
    is_wall(x, y, d) {
        var [x, y, z, d] = Maze.uniquify(x, y, d)
        if (!this.is_inside_of_field(x, y, z))
            return true;
        return this.walls[this.get_wall_index(x, y, z)];
    }
    is_known(x, y, d) {
        var [x, y, z, d] = Maze.uniquify(x, y, d)
        if (!this.is_inside_of_field(x, y, z))
            return true;
        return this.knowns[this.get_wall_index(x, y, z)];
    }
    update_wall(x, y, d, new_wall_state) {
        this.set_wall(x, y, d, new_wall_state);
        this.set_known(x, y, d, true);
    }
    set_wall(x, y, d, new_state) {
        var [x, y, z, d] = Maze.uniquify(x, y, d)
        if (!this.is_inside_of_field(x, y, z))
            return true;
        return this.walls[this.get_wall_index(x, y, z)] = new_state;
    }
    set_known(x, y, d, new_state) {
        var [x, y, z, d] = Maze.uniquify(x, y, d)
        if (!this.is_inside_of_field(x, y, z))
            return true;
        return this.knowns[this.get_wall_index(x, y, z)] = new_state;
    }
    get_maze_string() {
        let res = "";
        let s = this.size;
        for (let y = s - 1; y >= -1; --y) {
            // horizontal walls
            res += "+";
            for (let x = 0; x < s; ++x) {
                if (!this.is_known(x, y, Maze.North))
                    res += " . ";
                else if (this.is_wall(x, y, Maze.North))
                    res += "---";
                else
                    res += "   ";
                res += "+";
            }
            res += "\n";
            // vertical walls
            if (y === -1) break;
            res += "|";
            for (let x = 0; x < s; ++x) {
                // cell space
                if (JSON.stringify(this.start) === JSON.stringify([x, y]))
                    res += " S ";
                else if (this.goals.map(p => JSON.stringify(p)).includes(JSON.stringify([x, y])))
                    res += " G ";
                else
                    res += "   ";
                // vertical wall
                if (!this.is_known(x, y, Maze.East))
                    res += ".";
                else if (this.is_wall(x, y, Maze.East))
                    res += "|";
                else
                    res += " ";
            }
            res += "\n";
        }
        return res;
    }
    get_c_array_string({
        name = "maze_data",
        type = "const uint8_t",
        bit_order = [0, 1, 2, 3],
        y_origin_is_top = false,
    } = {}) {
        let s = this.size;
        let y_comment = y_origin_is_top ? "N-1-y" : "y";
        let res = `${type} ${name}[ /* ${y_comment} */ ${s}][ /* x */ ${s}] = {\n`;
        let y_array = [...Array(s).keys()]; // 0, 1, ... , s-1
        if (y_origin_is_top) y_array = y_array.reverse();
        for (let y of y_array) {
            res += "    {";
            for (let x = 0; x < s; ++x) {
                let hex = 0;
                if (this.is_wall(x, y, Maze.East)) hex += 1 << bit_order[0];
                if (this.is_wall(x, y, Maze.North)) hex += 1 << bit_order[1];
                if (this.is_wall(x, y, Maze.West)) hex += 1 << bit_order[2];
                if (this.is_wall(x, y, Maze.South)) hex += 1 << bit_order[3];
                res += `0x${hex.toString(16)}`;
                if (x < s - 1) res += ", ";
            }
            res += "},\n";
        }
        res += "};\n";
        return res;
    }
    static parse_maze_string(maze_string) {
        maze_string = maze_string.trim();
        let lines = maze_string.split(/\r?\n/);
        let s_x = Math.floor(lines[0].length / 4);
        let s_y = Math.floor(lines.length / 2);
        let s = Math.max(s_y, s_x);
        let maze = new Maze(s);
        lines = lines.reverse();
        for (let y = 0; y < s_y; y++) {
            // horizontal walls
            var line = lines[y * 2 + 0].trim();
            for (let x = 0; x < s_x; x++) {
                if (line[x * 4 + 2] === '-')
                    maze.update_wall(x, y, Maze.South, true);
                else if (line[x * 4 + 2] === ' ')
                    maze.update_wall(x, y, Maze.South, false);
            }
            // vertical walls
            var line = lines[y * 2 + 1].trim();
            for (let x = 0; x < s_x; x++) {
                if (line[x * 4 + 0] === '|')
                    maze.update_wall(x, y, Maze.West, true);
                else if (line[x * 4 + 0] === ' ')
                    maze.update_wall(x, y, Maze.West, false);
                if (line[x * 4 + 2] === 'S')
                    maze.start = [x, y];
                else if (line[x * 4 + 2] === 'G')
                    maze.goals.push([x, y]);
            }
        }
        return maze;
    }
    get_wall_index_from_maze_string_index(index) {
        let s = this.size;
        let y_span = (4 * s + 2) * 2;
        let y = Math.floor(s - index / y_span);
        let x = Math.floor(((index - 1 - 1) % (y_span / 2)) / 4);
        let z = Math.floor(index / (y_span / 2) + 1) % 2;
        // console.log(x, y, z);
        if (z === 0 && Math.floor((index) % (y_span / 2)) % 4 !== 2)
            return [x, y, Maze.East];
        else if (z === 1 && Math.floor(index % (y_span / 2)) % 4 !== 0)
            return [x, y, Maze.North];
        return [-1, -1, Maze.East];
    }
    toString() {
        let res = "";
        res += "size: " + this.size + "\n";
        res += "start: " + [this.start].map(p => "(" + p[0] + "," + p[1] + ")").join(",") + "\n";
        res += "goals: " + this.goals.map(p => "(" + p[0] + "," + p[1] + ")").join(",") + "\n";
        return res;
    }
}
